### PR TITLE
Refactor(dsolve): nth order reducible

### PR DESCRIPTION
**Pull request by Mohitbalwani26**
#### References to other Issues or PRs

SEE #18348

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/solvers/ode/tests/test_single.py [hunk 1]
```diff
@@ -501,16 +501,14 @@ def test_nth_linear_constant_coeff_undetermined_coefficients():
...
-    assert F(f(x)*D(f(x), x) + D(f(x), x, 2)) is None
-    assert F(D(x*f(y), y, 2) + D(u*y*f(x), x, 3)) is None  # no simplification by design
-    assert F(D(f(y), y, 2) + D(f(y), y, 3) + D(f(x), x, 4)) is None
-    assert F(D(f(x), x, 2) + D(f(x), x, 3)) == dict(n=2)
+    assert F(SingleODEProblem(D(y*f(x), x, y) + D(f(x), x), f(x), x)) == False
+    assert F(SingleODEProblem(D(y*f(y), y, y) + D(f(y), y), f(x), x)) == False
+    assert F(SingleODEProblem(f(x)*D(f(x), x) + D(f(x), x, 2), f(x), x))== False
+    assert F(SingleODEProblem(D(x*f(y), y, 2) + D(u*y*f(x), x, 3), f(x), x)) == False  # no simplification by design
+    assert F(SingleODEProblem(D(f(y), y, 2) + D(f(y), y, 3) + D(f(x), x, 4), f(x), x)) == False
+    assert F(SingleODEProblem(D(f(x), x, 2) + D(f(x), x, 3), f(x), x)) == True
```
Comment (oscarbenjamin) [inline sympy/solvers/ode/tests/test_single.py:511]: Maybe just change the function `F` rather than adding the same to all examples

**Comment by oscarbenjamin** [comment]
This is another example of a solver that is really just a substitution.

File: sympy/solvers/ode/tests/test_single.py (hunk 1 above)
Comment (Mohitbalwani26) [reply sympy/solvers/ode/tests/test_single.py:511]: Sorry, I couldn't understand the part change the function F?

Comment (oscarbenjamin) [reply sympy/solvers/ode/tests/test_single.py:511]: The function `F` is supposed to handle the repetitive parts in these test examples. You've just added repetitive changes to each example when you could have just changed the function `F`.

Comment (Mohitbalwani26) [reply sympy/solvers/ode/tests/test_single.py:511]:
````F = lambda eq: NthOrderReducible(SingleODEProblem(eq, f(x) ,  x) )._matches()
D = Derivative
assert F(D(y*f(x), x, y) + D(f(x), x)) == False
````
Something like this?

**Comment by Mohitbalwani26** [comment]
> This is another example of a solver that is really just a substitution.

Yes, I agree. After moving everything we should come up with either some generalised substitution function which try different substitutions or we can try to handle it inside pattern matching

**Comment by oscarbenjamin** [comment]
Looks good. Thanks

**Pull request merged**